src/hotspot/share/runtime/mutexLocker.hpp

```diff
@@ -254,6 +255,20 @@ class ConditionalMutexLocker: public MutexLockerImpl {
    }
 };
 
+// Mutex locker with NoSafePointVerifyer
+// At Safepoints locks can be broken. The verifier checks that there is no Safepoint in the scope
+// where the lock is held, and therefore the lock is not broken.
+class NoSafepointMutexLocker : public StackObj {
+  NoSafepointVerifier _nsv;
+  ConditionalMutexLocker _ml;
+
+public:
+  NoSafepointMutexLocker(Mutex* mutex, bool condition, Mutex::SafepointCheckFlag flag = Mutex::_safepoint_check_flag) :
+    _ml(mutex, condition, flag) {}
+  NoSafepointMutexLocker(Mutex* mutex, Mutex::SafepointCheckFlag flag = Mutex::_safepoint_check_flag) :
+    NoSafepointMutexLocker(mutex, true, flag) {}
+};
+
```

Comment:
Don't add this.  The locks that are no-safepoint-check-flags have implicit NoSafepointVerifier logic when you take them out.

Reply:
Ah ok. Thanks for the hint. I will look into this. It is a bit confusing what the locks do and do not do with SafePoints. Do they just not safepoint when trying to acquire the lock, or also verify that no safepoint is made while holding the lock?

Reply:
They verify that no safepoints happen while holding the lock.  There's a counter in JavaThread.

```
#ifdef ASSERT
  // Debug support for checking if code allows safepoints or not.
  // Safepoints in the VM can happen because of allocation, invoking a VM operation, or blocking on
  // mutex, or blocking on an object synchronizer (Java locking).
  // If _no_safepoint_count is non-zero, then an assertion failure will happen in any of
  // the above cases. The class NoSafepointVerifier is used to set this counter.
  int _no_safepoint_count;                             // If 0, thread allow a safepoint to happen
```

Reply:
I guess the comment should be updated to say we increment this when taking out a no-safepoint-check mutex.

Reply:
Ok, I removed the `NoSafepointVerifier`. It seems to work, and pass my verification that we are in a no_safepoint scope.